### PR TITLE
Directory Listing

### PR DESCRIPTION
**Pull request by kglebows**
## 1. Autoindexing (of sorts)
If entered an directory as address it will refer to default file named <directoryname>.html
image

works also without "/" at the end
image

## 2. If directory listing is ON and we try to enter directory that has no default file we trigger directory listing
image

## 3. If directory listing is OFF we throw error page
image

### CODE REVIEW

File: src/HTTPResponse.cpp [hunk 1]
```diff
@@ -56,16 +61,90 @@ void HTTPResponse::handleRequestGET(const HTTPRequest& request, const ServerConf
...
+			std::string content((std::istreambuf_iterator<char>(folderHtmlFile)), std::istreambuf_iterator<char>());
+			assignResponse(200, content, "text/html");
+			folderHtmlFile.close();
+			return;
+		}
+
+		/* if no default folder file found - directory listing */
+		if (serverConfig.directoryListing) {
+			DIR* dir = opendir(fullPath.c_str());
+			if (dir != nullptr) {
```
Comment (RealConrad) [inline src/HTTPResponse.cpp]: this is c++11

File: src/HTTPResponse.cpp [hunk 2]
```diff
@@ -56,16 +61,90 @@ void HTTPResponse::handleRequestGET(const HTTPRequest& request, const ServerConf
...
+				INFO("Serving Directory Listing of: " << fullPath);
+				struct dirent* entry;
+				std::string content = "<html><body><h1>Directory Listing of " + uri + "</h1><ul>";
+				while ((entry = readdir(dir)) != nullptr) {
```
Comment (RealConrad) [inline src/HTTPResponse.cpp]: c++11 keyword

File: src/HTTPResponse.cpp [hunk 3]
```diff
@@ -56,16 +61,90 @@ void HTTPResponse::handleRequestGET(const HTTPRequest& request, const ServerConf
...
+					if (strcmp(entry->d_name, ".") == 0)
```
Comment (RealConrad) [inline src/HTTPResponse.cpp]: Dont use C-style funtions, use cpp

File: src/HTTPResponse.cpp [hunk 4]
```diff
@@ -56,16 +61,90 @@ void HTTPResponse::handleRequestGET(const HTTPRequest& request, const ServerConf
 	}
 }
 
+std::string HTTPResponse::extractFolderName(const std::string& uri) {
+	if (uri.empty()) return "";
+	size_t lastSlashPos;
+	if (uri.back() == '/') {
```
Comment (RealConrad) [inline src/HTTPResponse.cpp]: `.back()` is c++11

**Review by RealConrad — CHANGES_REQUESTED** [review]

File: src/SocketManager.cpp [hunk 5]
```diff
@@ -1,4 +1,5 @@
 #include "SocketManager.hpp"
+#include "signal.h"
```
Comment (RealConrad) [inline src/SocketManager.cpp]: Can you move this into the `.hpp` file? As thats where all of our other library includes are

File: src/HTTPResponse.cpp [hunk 6]
```diff
@@ -56,16 +61,91 @@ void HTTPResponse::handleRequestGET(const HTTPRequest& request, const ServerConf
...
+
 void HTTPResponse::serveFile(const ServerConfig& serverConfig, const std::string& uri) {
-	std::string filePath = serverConfig.rootDirectory + (uri == "/" ? "/index.html" : uri);
-	std::ifstream file(filePath.c_str());
-	INFO("Serving file: " << filePath);
-	if (file) {
-		std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
-		assignResponse(200, content, determineContentType(uri));
-		file.close();
+	std::string fullPath = serverConfig.rootDirectory + uri;
```
Comment (RealConrad) [inline src/HTTPResponse.cpp]: Can you break up this function into helper functions? The function is around 70 lines of code and is hard to read

**Review by RealConrad — CHANGES_REQUESTED** [review]

**Comment by kglebows** [comment]
Splitted function and added multiple slashes handler:
image

**Review by RealConrad — APPROVED** [review]

**Pull request merged**